akka-remote/src/main/scala/akka/remote/artery/BufferPool.scala

```diff
@@ -401,6 +443,30 @@ private[remote] final class EnvelopeBuffer(val byteBuffer: ByteBuffer) {
     }
   }
 
+  private def writeByteString(tagOffset: Int, bs: ByteString): Unit = {
+    val length = bs.length
+    if (length > 65535)
+      throw new IllegalArgumentException("ByteStrings longer than 65535 cannot be encoded in the envelope")
+
+    byteBuffer.putInt(tagOffset, byteBuffer.position())
+
+    if (length == 0) {
+      byteBuffer.putShort(0)
+    } else {
+      byteBuffer.putShort(length.toShort)
+      ensureLiteralCharsLength(length)
+
+      @tailrec def doWriteByteString(bs: ByteString): Unit = {
+        bs match {
+          case cs: ByteString1C ⇒ byteBuffer.put(cs.asByteBuffer)
+          case _                ⇒ doWriteByteString(bs.compact)
+        }
```

Comment:
as we've benchmarked during HTTP work, compacting and retrying should give a good result here